src/org/jitsi/impl/neomedia/transform/fec/FlexFec03HeaderReader.java

```diff
@@ -0,0 +1,122 @@
+ * |k|                                                             |
+ * +-+                   Mask [46-108] (optional)                  |
+ * |                                                               |
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * |                     ... next in SSRC_i ...                    |
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ */
+public class FlexFec03HeaderReader
+{
+    /**
+     * The minimum size of the FlexFec header, in bytes.
+     * We include here one instance of SSRC_i and SN base_i, as well as
+     * the smallest possible packet mask.
+     */
+    private static int HEADER_MIN_SIZE_BYTES = 20;
+
+    /**
+     * The offset of the mask in the flexfec header, relative to the start
+     * of the flexfec header
+     */
+    private static int MASK_START_OFFSET_BYTES = 18;
+
+    /**
+     * Parse a buffer pointing to FlexFec data.
+     * @param flexFecPacket the (empty) FlexFec03Packet which the parsed
+     * values will be written into
+     * @param buffer buffer which contains the flexfec data
+     * @param flexFecOffset the flexFecOffset in buffer at which the flexfec header starts
+     * @param length length of the buffer
+     * @return true if parsing succeeded, false otherwise
+     */
+    public static boolean readFlexFecHeader(FlexFec03Packet flexFecPacket,
+                                            byte[] buffer, int flexFecOffset,
+                                            int length)
+    {
+        if (length < HEADER_MIN_SIZE_BYTES)
+        {
+            return false;
+        }
+        boolean retransmissionBit = ((buffer[flexFecOffset] & 0x80) >> 7) == 1;
+        if (retransmissionBit)
+        {
+            // We don't support flexfec retransmissions
+            return false;
+        }
+        int maskType = (buffer[flexFecOffset] & 0x40) >> 6;
+        if (maskType != 0)
+        {
+            // We only support flexible (f == 0) mask type
+            return false;
+        }
+
+        int ssrcCount = buffer[flexFecOffset + 8];
+        if (ssrcCount > 1)
+        {
+            // We only support a single protected ssrc
+            return false;
+        }
+
+        long protectedSsrc =
+            RTPUtils.readUint32AsLong(buffer, flexFecOffset + 12);
+        int seqNumBase =
+            RTPUtils.readUint16AsInt(buffer,flexFecOffset + 16);
+
+        FlexFec03Mask mask =
+            new FlexFec03Mask(buffer,
+                flexFecOffset + MASK_START_OFFSET_BYTES, seqNumBase);
+
+        // HEADER_MIN_SIZES_BYTES already includes the the size of the smallest
+        // possible packet mask, but maskSizeBytes will include that as well,
+        // so subtract the size of the smallest possible packet mask (2)
+        // here when we want to calculate the total.
+        int flexFecHeaderSize = HEADER_MIN_SIZE_BYTES - 2 + mask.lengthBytes();
```

Comment:
Do we need an extra check for the length here somewhere? We haven't checked that the buffer is long enough to contain a possible long mask.

Reply:
do you mean `buffer`, which is passed in? we only read from that, and we do verify that it's at least the minimum size (via the separate 'length' argument...this was somewhat of a port from chrome's code where the separate pointer/length is necessary...here using `buffer.length` is probably better).  but either way, we never write to that buffer, only read from it, and we don't write to a buffer in here.

Reply:
oh, i think i see what you mean...you mean after parsing the actual mask size, verifying the buffer matches it? unfortunately, we have to read most of the mask in order to even determine the size, but you're right that we don't read past the last k bit (for the large mask), so there is a check we could add there.  i'll add it.

Reply:
when looking at this i found that the way i'd set up the interaction between the packet and the header was a bit awkward and made testing this hard...i refactored that a bit and it feels better now (and also checks for this case)